investigate/fix badger async read/write issue

There might be a bug in the current master version of badger, that we use (which is the latest except one commit). The issue only becomes apparent when doing a lot of reads and writes at once (but still a reasonable amount of operations from what can be expected from a production server). In that case, the data that is just written, seems not be available instantly when doing a read. A few ms later however, it would be available.

I checked about this issue with the badger maintainers, and they think it sounds like a bug in their code, from what I am doing, so we'll need to create a minimal test case to proof this is indeed the case.

You can see the bug happening in the following branch: https://github.com/zero-os/0-stor/tree/badger-async-issues

The initial unit test that made me note that bug, and talk about it with the dgraph team, is the following: https://github.com/zero-os/0-stor/blob/e9747bba0d7732822b9a8b68326b8a67f757cce8/server/db/badger/badger_test.go#L288-L290 (the weird part is that if you comment those lines you'll see the unit tests pass, hence why I know that a few ms that data would suddenly become available, as https://github.com/zero-os/0-stor/blob/e9747bba0d7732822b9a8b68326b8a67f757cce8/server/db/badger/badger_test.go#L294-L302 does the same Get operation, where the data is available)

So it seems that badger fails to perform under pressure. It's certainly different from the expected behavior, that when a write function returns, the data is expected to be available for use, but it seems that this is not immediately the case.

I thought that was it,but than today I was enabling a test that was disabled since a few months now, for some reason, and it seems to also have an issue related to not finding a key, after it was just written. In this unit test (https://github.com/zero-os/0-stor/blob/e9747bba0d7732822b9a8b68326b8a67f757cce8/client/client_test.go#L290-L343) we also write and read on multiple goroutines at once. Funny enough it was documented their (in that unit test) that as soon as they have more than 32 goroutines writing and reading, that the unit test fails. As I think this is suffering from the same badger-related bug, that would mean the bug has been in badger for a longer time, which is hard to believe, but plausible. It certainly wouldn't be the first bug we suffer from in badger (yet only the second).

You can see the tests fail at travis, here: https://travis-ci.org/zero-os/0-stor/builds/322666411

```
--- FAIL: TestSetIncremented_Async (0.85s)
	Error Trace:	badger_test.go:289
	Error:      	Received unexpected error:
	            	key not found
	Messages:   	key: d:82:waelNsQgdnXR-DA
```

```
time="2017-12-28T20:45:32Z" level=error msg="failed to read \"\\x00\\x00\\x00\\x00\" from given shard \"\\x00\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x03\\x00\\x00\\x00\" from given shard \"\\x03\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x00\\x00\\x00\\x00\" from given shard \"\\x00\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x04\\x00\\x00\\x00\" from given shard \"\\x04\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x02\\x00\\x00\\x00\" from given shard \"\\x02\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x01\\x00\\x00\\x00\" from given shard \"\\x01\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x06\\x00\\x00\\x00\" from given shard \"\\x06\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x06\\x00\\x00\\x00\" from given shard \"\\x06\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x05\\x00\\x00\\x00\" from given shard \"\\x05\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x04\\x00\\x00\\x00\" from given shard \"\\x04\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\f\\x00\\x00\\x00\" from given shard \"\\f\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\a\\x00\\x00\\x00\" from given shard \"\\a\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x10\\x00\\x00\\x00\" from given shard \"\\x10\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"\\x1f\\x00\\x00\\x00\" from given shard \"\\x1f\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"'\\x00\\x00\\x00\" from given shard \"'\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"*\\x00\\x00\\x00\" from given shard \"*\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"$\\x00\\x00\\x00\" from given shard \"$\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \".\\x00\\x00\\x00\" from given shard \".\\x00\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \",\\x00\\x00\\x00\" from given shard \",\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"/\\x00\\x00\\x00\" from given shard \"/\\x00\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"+\\x00\\x00\\x00\" from given shard \"+\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \")\\x00\\x00\\x00\" from given shard \")\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \">\\x00\\x00\\x00\" from given shard \">\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"E\\x00\\x00\\x00\" from given shard \"E\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"D\\x00\\x00\\x00\" from given shard \"D\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"B\\x00\\x00\\x00\" from given shard \"B\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"4\\x00\\x00\\x00\" from given shard \"4\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"F\\x00\\x00\\x00\" from given shard \"F\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"C\\x00\\x00\\x00\" from given shard \"C\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \";\\x00\\x00\\x00\" from given shard \";\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \":\\x00\\x00\\x00\" from given shard \":\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"8\\x00\\x00\\x00\" from given shard \"8\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"H\\x00\\x00\\x00\" from given shard \"H\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"I\\x00\\x00\\x00\" from given shard \"I\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"5\\x00\\x00\\x00\" from given shard \"5\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"G\\x00\\x00\\x00\" from given shard \"G\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"O\\x00\\x00\\x00\" from given shard \"O\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"G\\x00\\x00\\x00\" from given shard \"G\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"T\\x00\\x00\\x00\" from given shard \"T\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"P\\x00\\x00\\x00\" from given shard \"P\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"Q\\x00\\x00\\x00\" from given shard \"Q\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"H\\x00\\x00\\x00\" from given shard \"H\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"S\\x00\\x00\\x00\" from given shard \"S\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"U\\x00\\x00\\x00\" from given shard \"U\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"M\\x00\\x00\\x00\" from given shard \"M\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:33Z" level=error msg="failed to read \"R\\x00\\x00\\x00\" from given shard \"R\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"b\\x00\\x00\\x00\" from given shard \"b\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\\\\\x00\\x00\\x00\" from given shard \"\\\\\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"a\\x00\\x00\\x00\" from given shard \"a\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"d\\x00\\x00\\x00\" from given shard \"d\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"c\\x00\\x00\\x00\" from given shard \"c\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"]\\x00\\x00\\x00\" from given shard \"]\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"e\\x00\\x00\\x00\" from given shard \"e\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"`\\x00\\x00\\x00\" from given shard \"`\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"Z\\x00\\x00\\x00\" from given shard \"Z\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"q\\x00\\x00\\x00\" from given shard \"q\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"l\\x00\\x00\\x00\" from given shard \"l\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"k\\x00\\x00\\x00\" from given shard \"k\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"R\\x00\\x00\\x00\" from given shard \"R\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\\\\\x00\\x00\\x00\" from given shard \"\\\\\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"]\\x00\\x00\\x00\" from given shard \"]\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"S\\x00\\x00\\x00\" from given shard \"S\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"^\\x00\\x00\\x00\" from given shard \"^\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"y\\x00\\x00\\x00\" from given shard \"y\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"z\\x00\\x00\\x00\" from given shard \"z\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"x\\x00\\x00\\x00\" from given shard \"x\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"m\\x00\\x00\\x00\" from given shard \"m\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"z\\x00\\x00\\x00\" from given shard \"z\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"i\\x00\\x00\\x00\" from given shard \"i\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"|\\x00\\x00\\x00\" from given shard \"|\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"|\\x00\\x00\\x00\" from given shard \"|\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"~\\x00\\x00\\x00\" from given shard \"~\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\u007f\\x00\\x00\\x00\" from given shard \"\\u007f\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x80\\x01\\x00\\x00\" from given shard \"\\x80\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"q\\x00\\x00\\x00\" from given shard \"q\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"|\\x00\\x00\\x00\" from given shard \"|\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"{\\x00\\x00\\x00\" from given shard \"{\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"p\\x00\\x00\\x00\" from given shard \"p\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"}\\x00\\x00\\x00\" from given shard \"}\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x8c\\x01\\x00\\x00\" from given shard \"\\x8c\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x8d\\x01\\x00\\x00\" from given shard \"\\x8d\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x80\\x01\\x00\\x00\" from given shard \"\\x80\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x8b\\x01\\x00\\x00\" from given shard \"\\x8b\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x90\\x01\\x00\\x00\" from given shard \"\\x90\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x91\\x01\\x00\\x00\" from given shard \"\\x91\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x87\\x01\\x00\\x00\" from given shard \"\\x87\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x8f\\x01\\x00\\x00\" from given shard \"\\x8f\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x9f\\x01\\x00\\x00\" from given shard \"\\x9f\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xa2\\x01\\x00\\x00\" from given shard \"\\xa2\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x9d\\x01\\x00\\x00\" from given shard \"\\x9d\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xa3\\x01\\x00\\x00\" from given shard \"\\xa3\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xa1\\x01\\x00\\x00\" from given shard \"\\xa1\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xa6\\x01\\x00\\x00\" from given shard \"\\xa6\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xa5\\x01\\x00\\x00\" from given shard \"\\xa5\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xa9\\x01\\x00\\x00\" from given shard \"\\xa9\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x9d\\x01\\x00\\x00\" from given shard \"\\x9d\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x9c\\x01\\x00\\x00\" from given shard \"\\x9c\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x95\\x01\\x00\\x00\" from given shard \"\\x95\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xac\\x01\\x00\\x00\" from given shard \"\\xac\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x9f\\x01\\x00\\x00\" from given shard \"\\x9f\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xb2\\x01\\x00\\x00\" from given shard \"\\xb2\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xb3\\x01\\x00\\x00\" from given shard \"\\xb3\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xba\\x01\\x00\\x00\" from given shard \"\\xba\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xbd\\x01\\x00\\x00\" from given shard \"\\xbd\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xc2\\x01\\x00\\x00\" from given shard \"\\xc2\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xc1\\x01\\x00\\x00\" from given shard \"\\xc1\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xbb\\x01\\x00\\x00\" from given shard \"\\xbb\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xb9\\x01\\x00\\x00\" from given shard \"\\xb9\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xc2\\x01\\x00\\x00\" from given shard \"\\xc2\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xb5\\x01\\x00\\x00\" from given shard \"\\xb5\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xb7\\x01\\x00\\x00\" from given shard \"\\xb7\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xbf\\x01\\x00\\x00\" from given shard \"\\xbf\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xc8\\x01\\x00\\x00\" from given shard \"\\xc8\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xcc\\x01\\x00\\x00\" from given shard \"\\xcc\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xc2\\x01\\x00\\x00\" from given shard \"\\xc2\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xcd\\x01\\x00\\x00\" from given shard \"\\xcd\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xe5\\x01\\x00\\x00\" from given shard \"\\xe5\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xc0\\x01\\x00\\x00\" from given shard \"\\xc0\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xe8\\x01\\x00\\x00\" from given shard \"\\xe8\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xe5\\x01\\x00\\x00\" from given shard \"\\xe5\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xde\\x01\\x00\\x00\" from given shard \"\\xde\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xe7\\x01\\x00\\x00\" from given shard \"\\xe7\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xe8\\x01\\x00\\x00\" from given shard \"\\xe8\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xdf\\x01\\x00\\x00\" from given shard \"\\xdf\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xeb\\x01\\x00\\x00\" from given shard \"\\xeb\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xec\\x01\\x00\\x00\" from given shard \"\\xec\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xea\\x01\\x00\\x00\" from given shard \"\\xea\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xeb\\x01\\x00\\x00\" from given shard \"\\xeb\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x86\\x02\\x00\\x00\" from given shard \"\\x86\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xe5\\x01\\x00\\x00\" from given shard \"\\xe5\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xfc\\x01\\x00\\x00\" from given shard \"\\xfc\\x01\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xfd\\x01\\x00\\x00\" from given shard \"\\xfd\\x01\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xf1\\x01\\x00\\x00\" from given shard \"\\xf1\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xf0\\x01\\x00\\x00\" from given shard \"\\xf0\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xef\\x01\\x00\\x00\" from given shard \"\\xef\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xec\\x01\\x00\\x00\" from given shard \"\\xec\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xf5\\x01\\x00\\x00\" from given shard \"\\xf5\\x01\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xf4\\x01\\x00\\x00\" from given shard \"\\xf4\\x01\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x85\\x02\\x00\\x00\" from given shard \"\\x85\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x84\\x02\\x00\\x00\" from given shard \"\\x84\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x83\\x02\\x00\\x00\" from given shard \"\\x83\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x87\\x02\\x00\\x00\" from given shard \"\\x87\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x82\\x02\\x00\\x00\" from given shard \"\\x82\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x81\\x02\\x00\\x00\" from given shard \"\\x81\\x02\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xf4\\x01\\x00\\x00\" from given shard \"\\xf4\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\x81\\x02\\x00\\x00\" from given shard \"\\x81\\x02\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xff\\x01\\x00\\x00\" from given shard \"\\xff\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\xfc\\x01\\x00\\x00\" from given shard \"\\xfc\\x01\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
--- FAIL: TestConcurrentWriteRead (1.37s)
	client_test.go:327: Concurrent client: 1 - Size of the data: 128
	client_test.go:340: duration 47 ms
		
	client_test.go:327: Concurrent client: 1 - Size of the data: 512
	client_test.go:340: duration 119 ms
		
	client_test.go:327: Concurrent client: 2 - Size of the data: 128
	client_test.go:340: duration 43 ms
		
	client_test.go:327: Concurrent client: 2 - Size of the data: 512
	client_test.go:340: duration 122 ms
		
	client_test.go:327: Concurrent client: 4 - Size of the data: 128
	client_test.go:340: duration 44 ms
		
	client_test.go:327: Concurrent client: 4 - Size of the data: 512
	client_test.go:340: duration 135 ms
		
	client_test.go:327: Concurrent client: 8 - Size of the data: 128
	client_test.go:340: duration 68 ms
		
	client_test.go:327: Concurrent client: 8 - Size of the data: 512
	client_test.go:340: duration 138 ms
		
	client_test.go:327: Concurrent client: 16 - Size of the data: 128
	client_test.go:340: duration 41 ms
		
	client_test.go:327: Concurrent client: 16 - Size of the data: 512
	client_test.go:340: duration 140 ms
		
	client_test.go:327: Concurrent client: 32 - Size of the data: 128
	client_test.go:340: duration 40 ms
		
	client_test.go:327: Concurrent client: 32 - Size of the data: 512
	client_test.go:340: duration 141 ms
		
	client_test.go:327: Concurrent client: 64 - Size of the data: 128
	client_test.go:340: duration 36 ms
		
	client_test.go:327: Concurrent client: 64 - Size of the data: 512
	Error Trace:	client_test.go:318
			client_test.go:335
			asm_amd64.s:2337
	Error:      	Received unexpected error:
	            	(too many?) shards are unavailable
	Messages:   	fail read data
	client_test.go:340: duration 142 ms
		
time="2017-12-28T20:45:34Z" level=error msg="failed to read \"\\b\\x00\\x00\\x00\" from given shard \"\\b\\x00\\x00\\x00\": rpc error: code = Unavailable desc = transport is closing"
time="2017-12-28T20:45:37Z" level=error msg="failed to read \"\\r\\x00\\x00\\x00\" from given shard \"\\r\\x00\\x00\\x00\": zstordb: key is no found"
time="2017-12-28T20:45:37Z" level=error msg="error while closing listed shard (127.0.0.1:43168): grpc: the client connection is closing"
time="2017-12-28T20:45:37Z" level=error msg="error while closing listed shard (127.0.0.1:43168): grpc: the client connection is closing"
FAIL
coverage: 81.4% of statements
FAIL	github.com/zero-os/0-stor/client	12.611s
```

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -65,7 +65,7 @@
   branch = "master"
   name = "github.com/dgraph-io/badger"
   packages = [".","options","protos","skl","table","y"]
-  revision = "9e707c0f2074afbd43bcd11d058fed6ca7609105"
+  revision = "05c461659d9dd8579a5d26175fbb454e9dca6cf1"
 
 [[projects]]
   name = "github.com/dgrijalva/jwt-go"

--- a/client/client_test.go
+++ b/client/client_test.go
@@ -304,8 +304,6 @@ func TestMultipleDownload_Issue208(t *testing.T) {
 }
 
 func TestConcurrentWriteRead(t *testing.T) {
-	t.SkipNow()
-
 	servers, serverClean := testGRPCServer(t, 4)
 	defer serverClean()
 

--- a/server/api/grpc/object_api.go
+++ b/server/api/grpc/object_api.go
@@ -211,8 +211,14 @@ func (api *ObjectAPI) ListObjectKeys(req *pb.ListObjectKeysRequest, stream pb.Ob
 		for item := range ch {
 			// copy key to take ownership over it
 			key = item.Key()
-			if len(key) <= scopeKeyLength {
-				panic("invalid item key (filtered key is too short)")
+			if n := len(key); n < scopeKeyLength {
+				panic("invalid item key '" + string(key) +
+					"' (filtered key is too short)")
+			} else if n == scopeKeyLength {
+				log.Warningf(
+					"skipping listed key result, '%s', as it equals the given scopeKey",
+					scopeKey)
+				continue
 			}
 			key = key[scopeKeyLength:]
 			resp.Key = make([]byte, len(key))

--- a/server/db/badger/badger_test.go
+++ b/server/db/badger/badger_test.go
@@ -301,14 +301,9 @@ func TestSetIncremented_Async(t *testing.T) {
 		_, ok = m[output.Key]
 		require.False(t, ok)
 
-		// TODO: UN-COMMENT
-		// For some reason this line returns ErrKeyNotFound,
-		// even though this key is just added, not sure if this a bug
-		// in our code, a bug in badger, or simply to be expected.
-		// Either way, some ms later, it is available, as can be see in the for loop, below.
-		/*value, err := ddb.Get([]byte(output.Key))
+		value, err := ddb.Get([]byte(output.Key))
 		require.NoErrorf(t, err, "key: %s", output.Key)
-		require.Equal(t, output.Value, string(value))*/
+		require.Equal(t, output.Value, string(value))
 
 		m[output.Key] = output.Value
 	}

--- a/server/db/memory/memory_test.go
+++ b/server/db/memory/memory_test.go
@@ -285,14 +285,9 @@ func TestSetIncremented_Async(t *testing.T) {
 		_, ok = m[output.Key]
 		require.False(t, ok)
 
-		// TODO: UN-COMMENT
-		// For some reason this line returns ErrKeyNotFound,
-		// even though this key is just added, not sure if this a bug
-		// in our code, a bug in badger, or simply to be expected.
-		// Either way, some ms later, it is available, as can be see in the for loop, below.
-		/*value, err := mdb.Get([]byte(output.Key))
+		value, err := mdb.Get([]byte(output.Key))
 		require.NoErrorf(t, err, "key: %s", output.Key)
-		require.Equal(t, output.Value, string(value))*/
+		require.Equal(t, output.Value, string(value))
 
 		m[output.Key] = output.Value
 	}

